cdc/model/changefeed.go

```diff
@@ -305,6 +310,55 @@ func (info *ChangeFeedInfo) fixState() {
 	}
 }
 
+// fixSinkProtocol attempts to fix protocol incompatible.
+// We no longer support the acceptance of protocols that are not known.
+// The ones that were already accepted need to be fixed.
+func (info *ChangeFeedInfo) fixSinkProtocol() {
+	sinkURIParsed, err := url.Parse(info.SinkURI)
+	if err != nil {
+		log.Warn("parse sink URI failed", zap.Any("sink URI", info.SinkURI), zap.Error(err))
+		// SAFETY: It is safe to ignore this unresolvable sink URI here,
+		// as it is almost impossible for this to happen.
+		// If we ignore it when fixing it after it happens,
+		// it will expose the problem when starting the changefeed,
+		// which is easier to troubleshoot than reporting the error directly in the bootstrap process.
+		return
+	}
+	rawQuery := sinkURIParsed.Query()
+	protocolStr := rawQuery.Get(config.ProtocolKey)
+
+	needsFix := func(protocolStr string) bool {
+		var protocol config.Protocol
+		err = protocol.FromString(protocolStr)
+		// There are two cases:
+		// 1. there is an error indicating that the old ticdc accepts
+		//    a protocol that is not known. It needs to be fixed as open protocol.
+		// 2. If it is default, then it needs to be fixed as open protocol.
+		return err != nil || protocolStr == config.ProtocolDefault.String()
+	}
+
+	openProtocolStr := config.ProtocolOpen.String()
+	// The sinkURI always has a higher priority.
+	if protocolStr != "" {
+		if needsFix(protocolStr) {
+			rawQuery.Set(config.ProtocolKey, openProtocolStr)
+			sinkURIParsed.RawQuery = rawQuery.Encode()
+			fixedSinkURI := sinkURIParsed.String()
+			log.Info("handle incompatible protocol from sink URI",
+				zap.String("old URI", info.SinkURI),
+				zap.String("fixed URI", fixedSinkURI))
```

Comment:
Could you hide username/password?

Reply:
Good catch! Now it outputs only the insensitive parts that we may modify.